File: native/android/app/src/main/java/org/membercovenantpath/viewer/data/AppPrefs.kt
package org.membercovenantpath.viewer.data

import android.content.Context
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map

/** Theme preference (mirrors the Flutter ThemeController: system → light → dark cycle). */
enum class ThemeChoice { System, Light, Dark;
    val label: String get() = name
}

private val Context.dataStore by preferencesDataStore(name = "covenant_path_prefs")

/**
 * Persisted UI/account prefs via DataStore (the native analog of shared_preferences):
 *  • theme mode (theme/app_theme.dart ThemeController),
 *  • biometric app-lock toggle (biometric_gate.dart),
 *  • current stake id (dashboard_page.dart 'current_stake_id'),
 *  • one-time passkey-upsell flag (dashboard_page.dart 'passkey_suggested').
 */
class AppPrefs(private val context: Context) {

    private val themeKey = stringPreferencesKey("theme_mode")
    private val lockKey = booleanPreferencesKey("biometric_lock_enabled")
    private val stakeKey = stringPreferencesKey("current_stake_id")
    private val passkeySuggestedKey = booleanPreferencesKey("passkey_suggested")

    val theme: Flow<ThemeChoice> = context.dataStore.data.map { p ->
        when (p[themeKey]) {
            "light" -> ThemeChoice.Light
            "dark" -> ThemeChoice.Dark
            else -> ThemeChoice.System
        }
    }

    suspend fun setTheme(choice: ThemeChoice) {
        context.dataStore.edit { it[themeKey] = choice.name.lowercase() }
    }

    val appLock: Flow<Boolean> = context.dataStore.data.map { it[lockKey] ?: false }
    suspend fun setAppLock(on: Boolean) {
        context.dataStore.edit { it[lockKey] = on }
    }
    suspend fun appLockNow(): Boolean = appLock.first()

    val currentStakeId: Flow<String?> = context.dataStore.data.map { it[stakeKey] }
    suspend fun currentStakeIdNow(): String? = currentStakeId.first()
    suspend fun setCurrentStakeId(id: String) {
        context.dataStore.edit { it[stakeKey] = id }
    }

    suspend fun passkeySuggested(): Boolean = context.dataStore.data.map { it[passkeySuggestedKey] ?: false }.first()
    suspend fun markPasskeySuggested() {
        context.dataStore.edit { it[passkeySuggestedKey] = true }
    }
}

File: native/ios/Sources/CovenantPathKit/Services/AppPrefs.swift
import Foundation
import Observation

/// Small UserDefaults-backed preference store (port of the various SharedPreferences keys the
/// Flutter app uses: theme, biometric lock, remembered stake, the one-time passkey upsell flag).
public enum AppPrefs {
    private static let store = UserDefaults.standard

    // keys (match the Flutter pref keys where shared semantics matter)
    static let themeMode = "theme_mode"
    static let biometricLock = "biometric_lock_enabled"
    static let currentStake = "current_stake_id"
    static let passkeySuggested = "passkey_suggested"

    public static func string(_ key: String) -> String? { store.string(forKey: key) }
    public static func setString(_ key: String, _ value: String) { store.set(value, forKey: key) }
    public static func bool(_ key: String, default def: Bool = false) -> Bool {
        store.object(forKey: key) == nil ? def : store.bool(forKey: key)
    }
    public static func setBool(_ key: String, _ value: Bool) { store.set(value, forKey: key) }
}

/// The persisted appearance mode (system/light/dark), with a cycle for the single menu toggle.
/// Mirrors `theme/app_theme.dart` `ThemeController`. The view maps this to SwiftUI's `ColorScheme?`.
public enum AppThemeMode: String, CaseIterable, Sendable {
    case system, light, dark
    public var label: String {
        switch self {
        case .system: return "System"
        case .light: return "Light"
        case .dark: return "Dark"
        }
    }
}

@MainActor
@Observable
public final class ThemeController {
    public private(set) var mode: AppThemeMode

    public init() {
        let raw = AppPrefs.string(AppPrefs.themeMode)
        mode = AppThemeMode(rawValue: raw ?? "") ?? .system
    }

    public func set(_ m: AppThemeMode) {
        mode = m
        AppPrefs.setString(AppPrefs.themeMode, m.rawValue)
    }

    /// Cycle system → light → dark → system (port of `ThemeController.cycle`).
    public func cycle() {
        switch mode {
        case .system: set(.light)
        case .light: set(.dark)
        case .dark: set(.system)
        }
    }
}
